### DISCUSSION TITLE
如何理解最终hook到的方法呢？

### DISCUSSION THREAD

**Discussion by xmutzlq** [bytedance/bhook#47] [Q&A] [answered]
I/bytehook_tag: >>>>> hooked. stub: 79fe19d580, status: 13, caller_path_name: /data/app/com.bytedance.android.bytehook.sample-pygIMku2BdKOooTOQlfWGg==/lib/arm64/libhookee.so, sym_name: open, new_func: 79f052f9b4, prev_func: 0, arg: 0

I/bytehook_tag: >>>>> hooked. stub: 79fe19d780, status: 13, caller_path_name: /data/app/com.bytedance.android.bytehook.sample-pygIMku2BdKOooTOQlfWGg==/lib/arm64/libhookee.so, sym_name: __open_real, new_func: 79f052fa70, prev_func: 0, arg: 0

I/bytehook_tag: >>>>> hooked. stub: 79fe19d800, status: 0, caller_path_name: /data/app/com.bytedance.android.bytehook.sample-pygIMku2BdKOooTOQlfWGg==/lib/arm64/libhookee.so, sym_name: __open_2, new_func: 79f052fb2c, prev_func: 7a87cae218, arg: 0

```
__BIONIC_FORTIFY_INLINE
int open(const char* const __pass_object_size pathname, int flags)
        __overloadable
        __clang_error_if(__open_modes_useful(flags), "'open' " __open_too_few_args_error) {
#if __ANDROID_API__ >= 17 && __BIONIC_FORTIFY_RUNTIME_CHECKS_ENABLED
    return __open_2(pathname, flags);
#else
    return __open_real(pathname, flags);
#endif
}
```
为什么是__open_2方法有效，而不直接是open方法呢？
我能hook hookee.c中的方法hookee_test吗？

**Comment by Mr-JingShi** [comment, marked as answer]
#  define __BIONIC_FORTIFY_INLINE static __inline__ __always_inline

因为这个地方被inline啊。

如果你想确确实实的调用open方法的话，可以这样int fd =  (&open)("/dev/null", O_RDWR);试试

**Reply by xmutzlq** [reply]
用的&open的方式可以的 老哥稳
请教下bhook能hook住android.so中AAssetManager_open的方法吗？
我在hookee.c中加入如下方法并注册
```
static void check_asset(JNIEnv *env, jobject thiz, jobject assetManager) {
  (void)thiz;
  AAssetManager *nativeasset = AAssetManager_fromJava(env, assetManager);
  AAsset *assetFile = AAssetManager_open(nativeasset, "AssetCheck.txt", AASSET_MODE_BUFFER);
  AAsset_close(assetFile);
}
```
/**
 * Open an asset.
 *
 * The object returned here should be freed by calling AAsset_close().
 */
AAsset* AAssetManager_open(AAssetManager* mgr, const char* filename, int mode);

最后状态返回13
I/bytehook_tag: >>>>> hooked. stub: 79ff61ff80, status: 13, caller_path_name: /system/lib64/libandroid.so, sym_name: AAssetManager_open, new_func: 79f042dbbc, prev_func: 0, arg: 0

**Reply by Mr-JingShi** [reply]
摘自native_manual.zh-CN.md：
ByteHook 是针对“调用者”执行 hook 的 hook 方案。比如 libart.so 调用了 libc.so 中的 malloc() 函数，当 hook malloc() 时，libart.so 是调用者（caller），libc.so 是被调用者（callee）。ByteHook 的 hook 行为是在修改调用者（libart.so）的 GOT 表。

老哥，估计是你把caller和callee搞错，你把caller_path_name填的是libandroid.so，其实应该填libhookee.so。
请参考下。

**Reply by xmutzlq** [reply]
嗯 对的，昨天仔细看了下确实是我理解错了
我的本意是想直接hook住AAssetManager_open而不是用过调用者，估计是要走inlinkhook的方式了
